### PR TITLE
Fix window counter (exclude hidden windows)

### PR DESCRIPTION
**Pull request by GeoffreyCoulaud**
Hide windows that hint not to be shown in the task bar.
Fix for #1629

### CODE REVIEW

**Review by slavicd — APPROVED** [review]
Ran in Fedora 35

**Comment by kanakshilledar** [comment]
nautilus window counter
Sorry, not working for me. I tried opening two nautilus windows and then tried copying one file to the other then after the transfer process ended the counter incremented by 2. I have attached a screenshot for reference.
tried logging out.

**Comment by GeoffreyCoulaud** [comment]
@kanakshilledar I could not reproduce, following your procedure I have only 2 nautilus indicators before, during and after the copy.  

Did you test this MR properly ?   
You need to disable the extension, build and install my branch, reload `gnome-shell` (`alt`+`f2`, `r`, `return`) and then try again. 

If it's still broken on your machine, I'll need to investigate deeper.

**Comment by slavicd** [comment]
I only needed the restart gnome shell part for the changes to take effect. I'll restate that the fix worked for me (I am, in fact, still running the code until the upstream merges this).

**Comment by kanakshilledar** [comment]
@GeoffreyCoulaud  I did the following
- Removed the old extension from gnome-extension using the `remove` option
- Cloned your branch with `$ git clone https://github.com/GeoffreyCoulaud/dash-to-dock`
- Installed by `$ make` and `$ make install`
- reloaded the shell using `alt` + `f2`, `r`)
But still I am having that issue.
I am having Gnome shell version 41.3 and Endeavour OS (arch linux)
I have also attached a sample video [video_link](https://youtu.be/BVQG-aAdhW0)

**Comment by kanakshilledar** [comment]
@GeoffreyCoulaud  I am using X11 and getting this issue.
```
$ loginctl show-session 7 -p Type
Type=x11 
```

**Comment by GeoffreyCoulaud** [comment]
Yup, I tested again and It's broken in both now. I'm reading the mutter docs to see what's going wrong.

**Comment by GeoffreyCoulaud** [comment]
@kanakshilledar You may want to report that to the EndeavourOS maintainers.  
I've checked with my Manjaro install, mutter is the same version so it's likely not that.  
My fix works on Arch, Manjaro and Fedora.

I've also tried filtering with the `get_window_type()` function and only keeping `NORMAL`, `DIALOG` and `MODAL_DIALOG` ([docs here](https://gjs-docs.gnome.org/meta9~9_api/meta.windowtype)) and even that is not working. Maybe it's somewhere earlier in the call stack that breaks ? I have no idea.

**Comment by kanakshilledar** [comment]
Oh okay I will try reporting them... Thanks for the suggestion

**Comment by kanakshilledar** [comment]
I tried Reborn OS and in the live image I added the extension and it was working fine only once I felt that issue. So maybe it's with Endeavour OS only

File: appIcons.js
Comment (3v1n0) [inline appIcons.js:1236]: This is not needed, it should be pretty known for who is into developing a shell extension :)

File: appIcons.js
Comment (3v1n0) [inline appIcons.js:1239]: I was thinking to include this for some time now too... I've been always wondering if this would break the case of showing the windows into the previews-menu though.

In fact while the windows may not be to be shown in the lists, *maybe* we may want to show them in the previews overview... But I'm not either sure about :)

**Review by 3v1n0 — CHANGES_REQUESTED** [review]
idea looks ok though, but I think it must be at upper level, so outside the `settings.get_boolean('isolate-workspaces')` check scope.

Ideally you can just `return windows.filter(w => !w.skipTaskbar);`

**Comment by 3v1n0** [comment]
Ah, and please remove the duplicated filter on `_rebuildMenu`.

**Comment by GeoffreyCoulaud** [comment]
Thanks for the review.  
I implemented the changes, I also added curly braces to the `isolate_monitors` check to be more consistent.  

About the ability to separate visibility in the indicator and in the preview, I'm not sure about that. I think having a different number of indicators and previews in unexpected for the user. Having options is great, still I believe the added complexity and the difference from the overview dock is not worth it.

File: appIcons.js
Comment (3v1n0) [inline appIcons.js:1235]: Since you're there, can you please change to `===` equality check?

File: appIcons.js
Comment (3v1n0) [inline appIcons.js:1241]:
```suggestion
    if (settings.get_boolean('isolate-monitors')) {
```

File: appIcons.js
Comment (GeoffreyCoulaud) [reply appIcons.js:1235]: Will do !

File: appIcons.js
Comment (3v1n0) [inline appIcons.js:1247]: I like using the property better here:

```suggestion
    return windows.filter(w => !w.skipTaskbar);
```

**Review by 3v1n0 — APPROVED** [review]

**Comment by FreddieChopin** [comment]
Sorry for noob question, but is the thing you discuss above the same as the issue I'm seeing right now:
Screenshot from 2022-01-31 15-20-45
I have just one Firefox window opened (I had two, but I closed it, yet there are 2 dots), only 1 window from e-mail client (sent the e-mail I was writing, so these additional windows are closed too), 1 windows of Nautilus.

**Comment by kanakshilledar** [comment]
I had the issue but it got fixed by the recent update.

**Comment by sergiotca** [comment]
> Sorry for noob question, but is the thing you discuss above the same as the issue I'm seeing right now: Screenshot from 2022-01-31 15-20-45 I have just one Firefox window opened (I had two, but I closed it, yet there are 2 dots), only 1 window from e-mail client (sent the e-mail I was writing, so these additional windows are closed too), 1 windows of Nautilus.

Yep, this is the issue at hand here.

**Comment by sergiotca** [comment]
@FreddieChopin Yep, that is the issue at hand here.

**Comment by sergiotca** [comment]
> I had the issue but it got fixed by the recent update.

What update are you talking about buddy?

**Comment by GeoffreyCoulaud** [comment]
@sergiotca Probably their distribution's package managers updated the extension.

To anyone unsure if the report applies to their bug : If it's about more indicators than should be, it's this issue and if you compile the extension from source you'll see it being resolved.

**Comment by tareqimbasher** [comment]
Thanks guys. When can we expect this fix to be pushed to extensions.gnome.org?

**Comment by kanakshilledar** [comment]
@sergiotca I manually compiled the Dash to Dock extension and the issue got resolved as suggested by @GeoffreyCoulaud.

**Pull request merged**
